pkg/cmd/server/admin/decrypt.go

```diff
@@ -0,0 +1,158 @@
+
+	# Decrypt with a prompted password:
+	$ %[1]s --in=secret3.data
+`
+
+func NewCommandDecrypt(commandName string, fullName, encryptFullName string, out io.Writer) *cobra.Command {
+	options := &DecryptOptions{Output: out}
+
+	cmd := &cobra.Command{
+		Use:     commandName,
+		Short:   fmt.Sprintf("Decrypt data encrypted with %q", encryptFullName),
+		Example: fmt.Sprintf(decryptExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Validate(args))
+			kcmdutil.CheckErr(options.Complete(args))
+			kcmdutil.CheckErr(options.Decrypt())
+		},
+	}
+
+	flags := cmd.Flags()
+
+	flags.StringVar(&options.InFile, "in", options.InFile, fmt.Sprintf("File containing the data to decrypt, in the format written by %q.", encryptFullName))
+	flags.StringVar(&options.OutFile, "out", options.OutFile, "File to write the decrypted data to. Written to stdout if omitted.")
+
+	flags.StringVar(&options.KeyArg, "key", options.KeyArg, "The password to use to decrypt. Prompted for if omitted.")
+	flags.StringVar(&options.InKeyFile, "keyin", options.InKeyFile, fmt.Sprintf("The file to read the decrypting key from. Must be a PEM file in the format written by %q.", encryptFullName))
+
+	// autocompletion hints
+	cmd.MarkFlagFilename("in")
+	cmd.MarkFlagFilename("out")
+	cmd.MarkFlagFilename("keyin")
+
+	return cmd
+}
+
+func (o *DecryptOptions) Complete(args []string) error {
+	if o.Stderr == nil {
+		o.Stderr = os.Stderr
+	}
+	if o.Output == nil {
+		o.Output = os.Stdout
+	}
+	if o.Input == nil {
+		o.Input = os.Stdin
+	}
+
+	switch {
+	case len(o.InFile) > 0:
+		// Read data from file
+		block, ok, err := pemutil.BlockFromFile(o.InFile, configapi.SecretBlockType)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("%s does not contain a %q PEM block", o.InFile, configapi.SecretBlockType)
+		}
+		o.Data = block
+	case o.Data == nil:
+		return fmt.Errorf("no input data provided")
+	}
+
+	switch {
+	case len(o.InKeyFile) > 0:
+		// Read key from file
+		block, ok, err := pemutil.BlockFromFile(o.InKeyFile, configapi.KeyBlockType)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("%s does not contain a %q PEM block", o.InKeyFile, configapi.KeyBlockType)
+		}
+		o.Key = block.Bytes
+	case len(o.KeyArg) > 0:
+		o.Key = []byte(o.KeyArg)
+	case util.IsTerminalReader(o.Input):
+		// Read key from stdin with prompting
+		o.Key = []byte(util.PromptForPasswordString(o.Input, o.Stderr, "Password: "))
+	}
+
+	return nil
+}
+
+func (o *DecryptOptions) Validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("no arguments are supported")
+	}
+	if len(o.InFile) == 0 && o.Data == nil {
+		return errors.New("--in must be provided")
+	}
+
+	return nil
+}
+
+func (o *DecryptOptions) Decrypt() error {
+	if len(o.Key) == 0 {
+		return errors.New("no password provided")
+	}
+
+	plaintext, err := x509.DecryptPEMBlock(o.Data, o.Key)
```

Comment:
Oh, I just got here.  `Key` in this case is not a key for a keyfile, it's the actual key content itself?  I don't think I care about prompting for that.  Why not force them to use a regular file and make the piping case work?